networkit/cpp/randomization/CurveballImpl.h

```diff
@@ -0,0 +1,245 @@
+/*
+ * CurveballImpl.h
+ *
+ * Author: Hung Tran <htran@ae.cs.uni-frankfurt.de>
+ */
+#ifndef RANDOMIZATION_CURVEBALL_IMPL_H
+#define RANDOMIZATION_CURVEBALL_IMPL_H
+
+#include <cassert>
+#include <utility>
+
+#include "../Globals.h"
+#include "../graph/Graph.h"
+
+namespace NetworKit {
+namespace CurveballDetails {
+
+// Global Definitions
+using TradeDescriptor = std::pair<node, node>;
+using tradeid_t = node;
+using count = node;
+
+using trade_vector = std::vector<TradeDescriptor>;
+using nodepair_vector = std::vector <std::pair<node, node> >;
+
+constexpr node INVALID_NODE = std::numeric_limits<node>::max();
+constexpr count LISTROW_END = std::numeric_limits<count>::max();
+constexpr tradeid_t TRADELIST_END = std::numeric_limits<tradeid_t>::max();
+
+struct edge_t : public std::pair<node, node> {
+    edge_t() : std::pair<node, node>() {}
+
+    edge_t(const std::pair <node, node> &edge) : std::pair<node, node>(
+        edge) {}
+
+    edge_t(const node &v1, const node &v2) : std::pair<node, node>(v1, v2) {}
+
+    static edge_t invalid() {
+        return edge_t(INVALID_NODE, INVALID_NODE);
+    }
+
+    void normalize() {
+        if (first > second)
+            std::swap(first, second);
+    }
+
+    bool is_invalid() const {
+        return (*this == invalid());
+    }
+};
+
+
+class CurveballAdjacencyList {
```

Comment:
I guess this class is needed to preallocate adjacency lists if you know the right degree. Such a class could be useful in a more general context. Could this be added to the Graph class instead?

If there is not enough time until Monday to do this, consider it for the future.